Adding ability to compare git references to spack install

vsoch opened the pull request:
This will allow a user to (from anywhere a Spec is parsed including both name and version) refer to a git commit in lieu of 
a package version, and be able to make comparisons with releases in the history based on commits (or with other commits). We do this by way of:

 - Adding a property, is_commit, to a version, meaning I can always check if a version is a commit and then change some action.
 - Adding an attribute to the Version object which can lookup commits from a git repo and find the last known version before that commit, and the distance
 - Construct new Version comparators, which are tuples. For normal versions, they are unchanged. For commits with a previous version x.y.z, d commits away, the comparator is (x, y, z, '', d). For commits with no previous version, the comparator is ('', d) where d is the distance from the first commit in the repo.
 - Metadata on git commits is cached in the misc_cache, for quick lookup later.
 - Git repos are cached as bare repos in `~/.spack/git_repos`
 - In both caches, git repo urls are turned into file paths within the cache

If a commit cannot be found in the cached git repo, we fetch from the repo. If a commit is found in the cached metadata, we do not recompare to newly downloaded tags (assuming repo structure does not change). The cached metadata may be thrown out by using the `spack clean -m` option if you know the repo structure has changed in a way that invalidates existing entries. Future work will include automatic updates.

# Finding previous versions
Spack will search the repo for any tags that match the string of a version given by the `version` directive. Spack will also search for any tags that match `v + string` for any version string. Beyond that, Spack will search for tags that match a SEMVER regex (i.e., tags of the form x.y.z) and interpret those tags as valid versions as well. Future work will increase the breadth of tags understood by Spack

For each tag, Spack queries git to determine whether the tag is an ancestor of the commit in question or not. Spack then sorts the tags that are ancestors of the commit by commit-distance in the repo, and takes the nearest ancestor. The version represented by that tag is listed as the previous version for the commit.

Not all commits will find a previous version, depending on the package workflow. Future work may enable more tangential relationships between commits and versions to be discovered, but many commits in real world git repos require human knowledge to associate with a most recent previous version. Future work will also allow packages to specify commit/tag/version relationships manually for such situations.

# Version comparisons.
The empty string is a valid component of a Spack version tuple, and is in fact the lowest-valued component. It cannot be generated as part of any valid version. These two characteristics make it perfect for delineating previous versions from distances. For any version x.y.z, (x, y, z, '', _) will be less than any "real" version beginning x.y.z. This ensures that no distance from a release will cause the commit to be interpreted as "greater than" a version which is not an ancestor of it.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding ability to compare git references to spack install

This works by way of adding a property, is_commit, to a version, and then given that a version
is a commit, we return a GitFetcherStrategy to use it. In additional, we instruct
the version command to generate a lookup of commits that keeps track of the order
(for comparing between commits) and the previous and next spack version (for comparison
of a commit with a spack version string). If the commit does not have known releases before
it, then the previous is None and we cannot determine the relationship and return False.
The same is true if the commit does not have any known releases after it, although it
is unlikely to hit this case unless the user is asking for a version that has been
released but not added to spack.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
okay updated to try and parse any semantic versions found in git tags.

vsoch commented:
@becker33 the last test is still finishing but we can probably discuss it now anyway. Let me know if this is in the right direction of what we had in mind!

becker33 commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -889,10 +890,20 @@ def fetch(self):
             tty.debug('Already fetched {0}'.format(self.stage.source_path))
             return
 
+        self.clone(commit=self.commit)
```
I don't think we are guaranteed to have a commit here -- could be a tag or branch.

becker33 commented on lib/spack/spack/paths.py:
```diff
@@ -56,6 +57,10 @@
 monitor_path = os.path.join(reports_path, "monitor")
 
+# We cache repositories (git) in first, extracted metadata in second
+user_repos_cache_path = os.path.join(user_config_path, 'git_repos')
+user_repos_metadata_path = os.path.join(user_config_path, 'git_metadata')
```
Anything that we put in the home directory, we have to add an option for people to put it somewhere else. Things in the home directory drive our users insane when they try to operate separate testing and production environments. (I've learned by being "yelled" at).

vsoch replied:
We don't want more people yelling at anyone!

vsoch replied:
But I'm not sure where else I'm supposed to put it?

vsoch replied:
I agree but looking at paths.py, there is a bunch of other content that should go somewhere else, but the somewhere else isn't clear. There should be a separate PR that allows this upper config to be customized (or the addition of a separate metadata place). I think it's out of scope for here, unless there is a suggestion for another existing place within spack for the repos?

tgamblin replied:
I think the best place for this right now would be the misc_cache -- you can use: 

```
os.path.join(spack.caches.misc_cache_location(), "git_metadata")
```

`misc_cache` is already a config preference.  By default it's one directory down from where you're currently putting stuff.

vsoch replied:
I gave this a shot, but I think there is a bug with misc_cache:

```
In [14]: _misc_cache()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/Desktop/Code/vsoch-spack/lib/spack/spack/caches.py in <module>
----> 1 _misc_cache()

~/Desktop/Code/vsoch-spack/lib/spack/spack/caches.py in _misc_cache()
     32 
     33 def _misc_cache():
---> 34     path = misc_cache_location()
     35     return spack.util.file_cache.FileCache(path)
     36 

~/Desktop/Code/vsoch-spack/lib/spack/spack/caches.py in misc_cache_location()
     24     providers and for which packages provide which tags.
     25     """
---> 26     path = spack.config.get('config:misc_cache')
     27     if not path:
     28         path = os.path.join(spack.paths.user_config_path, 'cache')

AttributeError: module 'spack' has no attribute 'config'
```
It might be an import issue - I was trying it in the caches.py.

becker33 commented on lib/spack/spack/version.py:
```diff
@@ -276,10 +304,23 @@ def satisfies(self, other):
+            version = Version(self.commits[self.version]['prev_version'])
+        elif other.is_commit and self.commits:
+            other = Version(self.commits[other.version]['prev_version'])
+        else:
```
```suggestion
        # Translate commits to highest versions prior to the commit
        if self.is_commit and self.commits:
            version = Version(self.commits[self.version]['prev_version'])
        if other.is_commit and self.commits:
            other = Version(self.commits[other.version]['prev_version'])
        else:
```

vsoch replied:
Oh wait this is outdated too, I can't commit it.

becker33 commented on lib/spack/spack/version.py:
```diff
@@ -276,10 +304,23 @@ def satisfies(self, other):
+        elif other.is_commit and self.commits:
+            other = Version(self.commits[other.version]['prev_version'])
+        else:
+            version = self.version
```
```suggestion
            version = self.version
            other = other.version
```
and `other.version` -> `other` for the rest of the method

vsoch replied:
I'm having trouble finding where this is because the file is changed locally from these lines, I'll push changes soon and I'll ask if you can point it out again.

vsoch replied:
okay I think this is outdated - I am replacing this with the below!

becker33 commented on lib/spack/spack/version.py:
```diff
@@ -276,10 +304,23 @@ def satisfies(self, other):
+        nself = len(version)
         nother = len(other.version)
-        return nother <= nself and self.version[:nother] == other.version
+        return nother <= nself and version[:nother] == other.version
```
We probably need to also store the distance (in commits) from the previous version, so we can compare two commits that stem from the same version

becker33 commented on lib/spack/spack/version.py:
```diff
@@ -331,13 +372,46 @@ def __lt__(self, other):
+            if prev_version:
+                return self.version < Version(prev_version)
+            return False
+
```
Can't this all be handled in a single case if we just translate each version to the "previous version" beforehand?

vsoch replied:
I don't understand what you are saying, could you please explain this more?

becker33 commented on lib/spack/spack/version.py:
```diff
@@ -331,13 +372,46 @@ def __lt__(self, other):
+                return self.version == other.version
+
+            # If the other isn't a commit, we need to check the spack version
+            exact_version = self.commits[self.version]['exact_version']
```
Would we be able to make this simpler if instead of resolving a separate `prev_version` and `exact_version`, we resolve a commit version to `implicit_version` with

1. If it's exactly a known version, just store that as `implicit_version`
2. If it's x commits newer than the known version `a.b.c`, store `a.b.c.z.x` ('z' is the lowest possible segment, so it will be lower than any real version after `a.b.c`, and storing the commit number assures we get the commits in order)

vsoch replied:
I don't know, if you want to try that works for me.

vsoch replied:
Can we meet and chat more about this so I understand?

becker33 commented on lib/spack/spack/version.py:
```diff
@@ -401,6 +475,36 @@ def intersection(self, other):
+        be one and the same with the source cache.
+
+        Args:
+            version: the requested version, which should be a commit.
```
This argument appears to be missing

vsoch replied:
Oops should be changed to fetcher.

becker33 commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +990,152 @@ class VersionError(spack.error.SpackError):
+        self.data = {}
+
+    @property
+    def repository_uri(self):
```
```suggestion
    def repository_url(self):
```

vsoch replied:
This isn't a url, it's a uri, unique object identifier (uri)

tgamblin replied:
I think URI is right here, but this needs a docstring that says this is the identifier for repos that is used within the repo and metadata caches.  And an example in the docstring -- that'll make it clearer.

vsoch committed:
Merge branch 'develop' into add/github-version-comparison

vsoch committed:
fixing isort issues

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
ping @becker33 ! I thought this was time sensitive (opened 12 days ago, and I think we only had under 2 months?) so would it be possible to get reviewed soon?

vsoch commented:
@spackbot hello!

vsoch commented:
@spackbot re-run pipeline

tgamblin commented:
@spackbot re-run pipeline

tgamblin commented:
@spackbot re-run pipeline

tgamblin commented:
@spackbot hello

tgamblin commented:
@spackbot hello

tgamblin commented:
@spackbot re-run pipeline

vsoch commented:
@spackbot commands

becker33 committed:
use pkg.git over pkg.url for git fetches

vsoch committed:
ensuring that we strip .git from a repository url, otherwise the output folder != the expected name

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
ensure we do not compare versions if either the primary or other version is a commit, and we have not indexed commits yet

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
we cannot check for commit versions if the package does not have a git attribute

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
cannot use rstrip as it also strips other letters in the string, using re instead

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

becker33 committed:
generate commit lookups for dev specs when added to concretization

becker33 committed:
type safety for Version.__lt__ and Version.satisfies

becker33 committed:
user other.commits if it exists and self.commits does not

becker33 committed:
fall back on prev_version in Version.__contains__

becker33 committed:
fix undefined variable

becker33 committed:
move version lookup to post-parse pass

becker33 committed:
add line in fetch_strategy.py for flake

vsoch commented on lib/spack/spack/spec.py:
```diff
@@ -4524,6 +4524,21 @@ def do_parse(self):
+                pkg = spec.package
+                if hasattr(pkg, 'git'):
+                    import spack.fetch_strategy as fetch
+                    spec.version.generate_commit_lookup(
```
Just testing this locally - for large packages there will be a long pause that feels like a freeze for spack install package@commit. Maybe a log info here would be appropriate?
```suggestion
                    tty.info(f"Generating commit lookup for {pkg.name}...")
                    spec.version.generate_commit_lookup(
```

vsoch commented on lib/spack/spack/spec.py:
```diff
@@ -4524,6 +4524,21 @@ def do_parse(self):
             raise SpecParseError(e)
 
+        # Generate lookups for git-commit-based versions
+        for spec in specs:
```
Just thinking out loud - is it ever the case that the same package is in a list of specs (for different versions) and would there be redundant generation (or starting the run and seeing that the commits exist and just exiting). I guess that's not so bad, but I wish we could have something like:

```bash
for pkg in get_unique_packages(specs):
   ...

vsoch commented:
@becker33 I found a good list of contender packages, but when I add one to buildin.mock and then write a test that (seems to have) the appropriate fixins, I always get an error about some file missing for the build. Here is how I derived the list (and the start of it, package that aren't py- or r- and have a git, and smallest number dependencies):

```
#!/bin/python

# Find the number of dependencies for all packages!
import spack.repo
import spack.spec

pkgs = set(spack.repo.all_package_names(False))
counts = {}
for pkg in pkgs:
    # Nope to Python
    if "py-" in pkg or "r-" in pkg:
        continue
    spec = spack.spec.Spec(pkg)
    if not hasattr(spec.package, "git"):
        continue
    counts[spec.package.name] = len(spec.package.dependencies)
    
    
counts = {k: v for k, v in sorted(counts.items(), key=lambda item: item[1])}

"""
Good contenders!

{'hsf-cmaketools': 0,
 'libblastrampoline': 0,
 'goblin-hmc-sim': 0,
 'nvdimmsim': 0,
 'libshm': 0,
 'cityhash': 0,
 'cinch': 0,
 'cusz': 0,
 'libtomlc99': 0,
 'ramulator': 0,
 'libbeato': 0,
 'lizard': 0,
 'readfq': 0,
 'dlpack': 0,
 'linkphase3': 0,
 'sse2neon': 0,
 'mlhka': 0,
 'libwhich': 0,
 'dramsim2': 0,
 'libopts': 0,
 'stream': 0,
 'hotspot': 0,
 'fasd': 0,
 'bannergrab': 0,
 'libmonitor': 0,
 'dnsmap': 0,
 'aeskeyfind': 0,
 'capnproto': 0,
 'shadow': 0,
 'pipework': 0,
 'gemmlowp': 0,
 'nn-c': 0,
 'highwayhash': 0,
 'cub': 0,
 'tkrzw': 0,
 'netkit-ftp': 0,
 'mshadow': 0,
 'sonlib': 0,
 'gdrcopy': 0,
 'macfuse': 0,
 'opennurbs': 0,
 'hbm-dramsim2': 0,
 'parsimonator': 0,
 'ppopen-at': 0,
 'veclibfort': 0,
 'csa-c': 0,
 'dnstracer': 0,
 'cntk1bitsgd': 0,
 'ninja': 1,
 'mcutils': 1,
 'amdblis': 1,
 'bamdst': 1,
 'dnstop': 1,
 'unqlite': 1,
 'ccs-qcd': 1,
 'pacbio-dazz-db': 1,
 'ppopen-appl-fdm-at': 1,
 'pxz': 1,
 'gtkorvo-dill': 1,
 'cohmm': 1,
 'libpng': 1,
 'occa': 1,
 'yajl': 1,
 'cnmem': 1,
 'lmbench': 1,
 'snap': 1,
 'mad-numdiff': 1,
 'hacckernels': 1,
 'talass': 1,
 'psm': 1,
 'taskflow': 1,
 'perfstubs': 1,
 'hapdip': 1,
 'blis': 1,
 'ocamlbuild': 1,
 'cppad': 1,
 'fastqvalidator': 1,
 'intel-llvm': 1,
 'amdlibflame': 1,
 'c-ares': 1,
 'ppopen-math-mp': 1,
 'bamaddrg': 1,
 'gpu-burn': 1,
 'brigand': 1,
 'amg': 1,
 'pegtl': 1,
 'umap': 1,
 'ftobjectlibrary': 1,
 'capstone': 1,
 'webbench': 1,
 'mpark-variant': 1,
 'yorick': 1,
 'xtl': 1,
 'graphmap': 1,
 'libxsmm': 1,
 'gotcha': 1,
 'mpilander': 1,
 'vvtest': 1,
 'r3d': 1,
 'ctre': 1,
 'intel-xed': 1,
 'ppopen-appl-dem-util': 1,
 'dmlc-core': 1,
 'miniaero': 1,
 'hpcg': 1,
 'gccxml': 1,
 'snbone': 1,
 'cppgsl': 1,
 'tycho2': 1,
 'gtkorvo-cercs-env': 1,
 'xhmm': 1,
 'stinger': 1,
 'sabre': 1,
 'ruby-narray': 1,
 'openblas': 1,
 'uchardet': 1,
 'wgsim': 1,
 'remhos': 1,
 'hapcut2': 1,
 'psmc': 1,
 'ember': 1,
 'tethex': 1,
 'phylobayesmpi': 1,
 'laghos': 1,
 'catch2': 1,
 'mdtest': 1,
 'saws': 1,
 'googletest': 1,
 'benchmark': 1,
 'cpu-features': 1,
 'nekbone': 1,
 'ruby-erubis': 1,
 'osqp': 1,
 'krims': 1,
 'autodock-gpu': 1,
 'cdbfasta': 1,
 'smartdenovo': 1,
 'cosp2': 1,
 'miniasm': 1,
 'nek5000': 1,
 'changa': 1,
 'miniamr': 1,
 'pacbio-damasker': 1,
 'mrnet': 1,
 'minivite': 1,
 'quicksilver': 1,
 'ppopen-math-vis': 1,
 'ebms': 1,
 'hub': 1,
 'blt': 1,
 'ragel': 1,
 'satsuma2': 1,
 'rocm-cmake': 1,
 'tldd': 1,
 'miniapp-ascent': 1,
 'flatcc': 1,
 'ppopen-appl-bem': 1,
 'kim-api': 1,
"""
```
and here is the simple test that fails:

```python
def test_install_commit(mock_packages, mock_archive, install_mockery, mock_fetch):
    """
    Test installing a git package from a commit.
    """
    install('libblastrampoline@c6c7bc5d4ae088bd7c519d58e3fb8b686d00db0c')
```
I tried it with three different ones and they all had the same outcome so I'm probably missing something about how to properly create a mock test package.

vsoch committed:
adding test for clone by command and supporting filesystem clone

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
style fixes

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
ensure we tell git who we are

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@spackbot run pipeline

vsoch commented:
@tgamblin looks like spackbot lost his creds?

tgamblin commented:
@spackbot run pipeline

tgamblin commented:
@spackbot run pipeline

tgamblin commented:
@spackbot run pipeline

tldahlgren commented:
Does this handle *any* commit (i.e., one in an active PR, not just in `develop`)?

tgamblin commented:
It should, if you do a `git fetch` on the right remote, but I don’t think we should preemptively try to do that if we don’t see a particular commit — we wouldn’t know which remote to fetch.

vsoch committed:
Adding comment about generating commit lookup

vsoch committed:
small updates to fetch strategy docs and version

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@becker33 just clone the branch and make the changes you suggested that I couldn't apply and push to the branch - I trust your work and it's much faster to go directly from your brain to the commit vs. using me as a middle person.

vsoch committed:
invalid syntax?

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
updating logic in commit parsing

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@becker33 I tried adding the changes that I wasn't sure about the lines for - let's meet offline and discuss the rest?

vsoch committed:
commits is never used

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
Can git repo change to have >= 5 branches?

becker33 committed:
refactor clone method signature

vsoch committed:
do not require that a fetcher only has one branch, tag, or commit

it is possible that we create a fetcher for a branch/tag, but then use a commit

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
Make sure that git log gets --all commits

becker33 committed:
try using tuples with empty string delimiter

becker33 committed:
flake

becker33 committed:
typo bugfix

becker33 committed:
fixup test for more robust git repo

becker33 committed:
make git commit install test stricter

becker33 committed:
test for comparing git versions

becker33 committed:
minor fixes to git version lookups

becker33 committed:
flake

becker33 committed:
isort

becker33 committed:
more isort

tgamblin committed:
fix isort/flake8 issues

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -870,6 +870,10 @@ def cachable(self):
 
+    @property
+    def source_digest(self):
+        return self.commit or self.tag or self.branch
```
This method isn't used anywhere -- delete

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -909,11 +923,12 @@ def fetch(self):
-                shutil.move(repo_name, self.stage.source_path)
+                if self.stage:
+                    self.stage.srcdir = repo_name
+                shutil.move(repo_name, dest)
```
If this doesn't end up using `os.rename`, the copy here will be very slow, and I think users will notice.  There should at least be a comment about this, and we should probably try to minimize the cost of this by cloning straight to `dest`.

`rsync` and even local `git clone` are ridiculously much faster than python copy.

becker33 replied:
I think this should be a separate PR

tgamblin replied:
Sounds good.

tgamblin replied:
Made an issue for this: https://github.com/spack/spack/issues/25930

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -950,14 +965,15 @@ def fetch(self):
-                shutil.move(repo_name, self.stage.source_path)
+                if self.stage:
+                    self.stage.srcdir = repo_name
+                shutil.move(repo_name, dest)
```
Same comment as above for this move.  The logic for whether we do a move or a direct clone should really be pulled out of here and handled in one place.

becker33 replied:
Same as above

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -895,10 +899,20 @@ def fetch(self):
+        Clone a repository to a path.
+
+        This is the fetch logic, but does not require a stage.
+        """
```
Arguments need documentation here.  Particularly `dest`

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1490,6 +1506,23 @@ def _from_merged_attrs(fetcher, pkg, version):
 
 
+def git_repo_for_package(pkg):
+    assert hasattr(pkg, 'git'), "Version lookups only allowed on git fetchers"
```
String here is not quite right. It's actually stronger than this -- we're requiring a package with a top-level `git` attribute.

A package could have 6 version directives, each with its own git URL, but we don't know how to compare those.  We're only considering packages with a canonical git repo from which everything is fetched.

Packages with directive-level `git` kwargs should probably be rewritten at some point to move to package-level `git` attributes

becker33 replied:
Removing entire method, see below.

tgamblin commented on lib/spack/spack/concretize.py:
```diff
@@ -80,6 +80,7 @@ def concretize_develop(self, spec):
             spec.variants.setdefault(
                 'dev_path', vt.SingleValuedVariant('dev_path', path))
             changed = True
+
```
not needed for this PR

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -1330,6 +1330,7 @@ def do_fetch(self, mirror_only=False):
 
         checksum = spack.config.get('config:checksum')
         fetch = self.stage.managed_by_spack
+
```
not needed for this PR

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1504,6 +1537,15 @@ def for_package_version(pkg, version):
     if not isinstance(version, Version):
         version = Version(version)
 
+    # if it's a commit, we must use a GitURLFetcher
```
```suggestion
    # if it's a commit, we must use a GitFetchStrategy
```

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1490,6 +1506,23 @@ def _from_merged_attrs(fetcher, pkg, version):
     return fetcher(**attrs)
 
 
+def git_repo_for_package(pkg):
```
needs a docstring

becker33 replied:
Removing entire method, see below.

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1490,6 +1506,23 @@ def _from_merged_attrs(fetcher, pkg, version):
+    match = re.search(repo_regex, pkg.git)
+    if not match:
+        tty.die("Cannot derive repository from %s." % pkg.git)
+    service, _, repository, _, name = match.groups()
```
This should not discard the username and the port from the original URL. It seems like this won't work on any internal repo URLs (bitbucket, gitlab, etc.)

becker33 replied:
I think this entire method is actually vestigial at this point, and we can just use `pkg.git` instead.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -38,6 +38,7 @@
 import spack.dependency
 import spack.directives
 import spack.error
+import spack.fetch_strategy
```
It doesn't look like this is used -- but it's not tripping flake8, so I'm a bit confused about it.  Why is it here?

becker33 replied:
I think that's vestigial from an old change. Not sure why flake8 didn't catch it...

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1657,6 +1658,7 @@ def _develop_specs_from_env(spec, env):
         spec.variants.setdefault(
             'dev_path', spack.variant.SingleValuedVariant('dev_path', path)
         )
+
```
remove random whitespace change.

tgamblin commented on lib/spack/spack/test/cmd/install.py:
```diff
@@ -225,7 +226,7 @@ def test_install_overwrite(
 
 def test_install_overwrite_not_installed(
-        mock_packages, mock_archive, mock_fetch, config, install_mockery
+        mock_packages, mock_archive, mock_fetch, config, install_mockery, monkeypatch
```
it looks like `monkeypatch` isn't used here.

tgamblin commented on lib/spack/spack/test/cmd/install.py:
```diff
@@ -30,9 +31,9 @@
-uninstall = SpackCommand('uninstall')
 buildcache = SpackCommand('buildcache')
 find = SpackCommand('find')
+uninstall = SpackCommand('uninstall')
```
moving this line seems unneeded.

tgamblin commented on lib/spack/spack/test/conftest.py:
```diff
@@ -60,6 +60,73 @@ def last_two_git_commits(scope='session'):
+    # 1. multiple branches
+    # 2. version tags on multiple branches
+    # 3. version order not equal to time order
+    with working_dir(repo_path):
```
can you diagram the repo in the docstring (like how the mock DB is documented - but with commits)?

This is what I got:

```
|
|  o fourth 1.x commit (1.2)
|  o third 1.x commit
|  |
o  | fourth main commit (v2.0)
o  | third main commit 
|  |
|  o second 1.x commit (v1.1)
|  o first 1.x commit
| /
|/
O second commit (v1.0)
o first commit
```

Ideally there would also be a 2.1 branch that is merged back into `main`.

tgamblin commented on lib/spack/spack/test/conftest.py:
```diff
@@ -60,6 +60,73 @@ def last_two_git_commits(scope='session'):
+        # Add two commits on main branch
+        write_file(filename, '[]')
+        git('add', filename)
+        git('commit', '-am', 'first commit')
```
don't need `-a` here since a) nothing is tracked yet and b) you added the file.

becker33 replied:
Ended up refactoring into a separate method, where it makes sense to take advantage of the fact that the `-a` flag does not harm anything.

tgamblin commented on lib/spack/spack/test/conftest.py:
```diff
@@ -60,6 +60,73 @@ def last_two_git_commits(scope='session'):
+        git('commit', '-am', 'fourth main commit')
+        git('tag', 'v2.0')
+
+        # Add two more commits on 1.x branch to ensure we aren't cheating by using time
```
git commits and unix timestamps have resolution of 1 second.  I suspect all of these commits have the same timestamp -- or at least it's possible for that to happen some of the time. 

You can use ISO 8601 dates in a `git commit` command:

```
git commit --date "2006-07-03 17:00:00 +0200" ...
git commit --date "2006-07-04 17:00:00 +0200" ...
git commit --date "2006-07-05 17:00:00 +0200" ...
```

etc.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
     """Raised for version checksum errors."""
+
+
+class CommitLookup(object):
```
docstring

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+        # Case 1: It's a path on the filesystem
+        if os.path.exists(self.fetcher.url) or self.fetcher.url.startswith("file://"):
+            return re.sub("file://", "", self.fetcher.url)
+        return re.sub("[.]git$", "", os.sep.join(self.fetcher.url.split(os.sep)[-3:]))
```
What's the -3 here?  Some git URLs have more than `hostname/project/org.git` -- they can be arbitrarily long.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+        Save the data to file
+        """
+        with open(self.repository_metadata_path, 'w') as fd:
+            sjson.dump(self.data, fd)
```
This should use a transactional move, and it should use locking to ensure that we don't get races -- use `spack.util.file_cache` like the other caches and it'll handle that automatically.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+
+    def get_commit_lookup(self, version, spack_versions):
+        """
+        Given a git handle, repository, and known spack versions, generate a lookup.
```
use paramter docstring format

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+            return self.data
+
+        # Were we given an already cloned path?
+        already_cloned = os.path.exists(self.repository_uri)
```
Make sure `os.path.join(self.repository_url, ".git")` is a directory; raise if not.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+
+        # But if we get here and no repostiory, this is a fail
+        if not os.path.exists(dest):
+            tty.die("There was an issue cloning %s to %s" % (self.fetcher.url, dest))
```
Is this needed? This should always be caught by `clone()` if we wrote `clone()` right.

Seems like clone errors should be caught and bundled up in a message like this (and raised if we're running `spack --debug`)

becker33 replied:
`clone` already raises all of its errors, this looks unreachable unless another process comes through and deletes `dest` between the two calls. I'll just delete this block.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+        for path in [dest, self.repository_metadata_path]:
+            parent = os.path.dirname(path)
+            if not os.path.exists(parent):
+                os.makedirs(parent)
```
Use `llnl.util.filesystem.mkdirp` as we use that in most of the code.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+        for tag in tags:
+            tag = tag.replace('refs/tags/', '', 1)
+            for v in versions:
+                if v in tag:
```
doesn't this match 1.1 to v1.1.1, v2.1.1, v3.1.1, etc?

becker33 replied:
... yes it would. And we don't have any good way to impose an ordering on the comparisons to help either.

I think for now I'll change to checking for the tag equaling the version or equaling 'v' + version.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+            tag = tag.replace('refs/tags/', '', 1)
+
+            # Try to derive a semantic version
+            match = SEMVER_REGEX.match(tag)
```
probably should document that we only handle undeclared versions if they are semvers.

becker33 replied:
If you have a good idea of how to broaden this, I think I would prefer that. But I'm not sure how to do it without getting spurious matches.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+        with working_dir(dest):
+            for spack_version, tag in tag_lookup.items():
+                command = ["rev-list", "-n", "1", "tags/%s" % tag]
+                commit = self.fetcher.git(*command, output=str)
```
Can avoid this git call by getting the commits along with the tag names when you call `git for-each-ref`:

```console
$ git for-each-ref --sort=creatordate --format "%(objectname) %(refname)" refs/tags
5829b44648f809a09d006b044d8244254a3d224a refs/tags/v0.8
6270fc3391b48cdbeeaebe9562bfe0a8c0b08f9e refs/tags/v0.8.5
1b388eadd4f6a50d98fa23c87480905c8b10a8f7 refs/tags/v0.8.10
5d0a9b7ce8f87612e1cbf580d67ecceed56cb0e7 refs/tags/v0.8.11
...
```

Then just parse the commits along with the tags.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+                            output=str, error=str).strip()
+                    ancestor_commits.append((tag_commit, int(distance)))
+                except spack.util.executable.ProcessError:
+                    # is-ancestor check will raise a ProcessError when False
```
would probably clearer to use fail_on_error=False for the call to merge-base, followed by a conditional

becker33 replied:
Used `ignore_errors` instead so we still raise on unexpected error codes.

tgamblin commented on lib/spack/spack/version.py:
```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+        # commits[0] is most recent, commits[-2] is first
+        commits = [c for c in commits.split('\n') if c]
+
+        for commit in commits:
```
This looks to me like it works, but it looks like it'll be very slow for large repositories -- I feel like there should be a way to do this that makes O(tags) calls to `git`, but it is not coming to me off the top of my head.

If this works then I say we get it in and optimize later -- would like to see more tests with more branch structure to ensure that (see comments above on the mock repo fixture)

tgamblin commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1504,6 +1537,15 @@ def for_package_version(pkg, version):
+    if version.is_commit and hasattr(pkg, "git"):
+        # Populate the version with comparisons to other commits
+        repository = git_repo_for_package(pkg)
+        fetcher = GitFetchStrategy(git=repository)
+        version.generate_commit_lookup(fetcher, pkg.versions)
```
All calls to `generate_commit_lookup` have to create a GitFetchStrategy -- I would just move the creation of the special fetch strategy into the `generate_commit_lookup` and pass the repository. I think that ends up being cleaner (it's really `generate_commit_lookup`'s fetcher).

tgamblin reviewed (changes requested):
I'm not entirely done reviewing this but I have a fair number of questions/change requests.

On other request is to take most of the git comparison logic and pull it out of `fetch_strategy.py` and `version.py`.  It really needs its own utility class.

becker33 committed:
refactor genreate_commit_lookup to instantiate its own fetcher

becker33 committed:
refactor git is-ancestor check to avoid try/except

becker33 committed:
refactor to avoid one git request per tag

becker33 committed:
fix version comparisons to tags

becker33 committed:
os.makedirs -> llnl.util.filesystem.mkdirp

becker33 committed:
remove unreachable code block

becker33 committed:
CommitLookup: refactor to lazy cache initialization for performance

becker33 committed:
improve error message and docstring

becker33 committed:
CommitLookup: use misc cache for locking

becker33 committed:
Merge branch 'add/github-version-comparison' of github.com:vsoch/spack into add/github-version-comparison

becker33 committed:
flake

becker33 committed:
bugfix: type(distance) must be int

becker33 committed:
ensure mock git version repo has distinct commit timestamps

becker33 committed:
docstring for mock_git_version_info

becker33 committed:
remove vestigial code from PR

becker33 committed:
improved docstrings and error messages

tgamblin committed:
add `spack.util.url.git_url_parse()` method and tests

Add a method to parse git URLs (but not arbitrary file paths). This method parses git
URLs into scheme, username, hostname, port, and path. It will raise an error if its
argument is not proper git URL.

- [x] add `spack.util.url.git_url_parse` method
- [x] add 31 tests for `git_url_parse`

becker33 committed:
Merge branch 'add/github-version-comparison' of github.com:vsoch/spack into add/github-version-comparison

becker33 committed:
refactor git parsing to make use of git_url_parse

tgamblin commented on lib/spack/spack/test/conftest.py:
```diff
@@ -60,6 +60,102 @@ def last_two_git_commits(scope='session'):
+def mock_git_version_info(tmpdir, scope="function"):
+    """Create a mock git repo with known structure
+
+    The structure of commits in this repo is as follows:
```
use two colons and indent so that text below is preformatted in docs. https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

tgamblin reviewed (changes requested):
This is looking great -- only final requests are:

1. Use bare repositories and `git fetch` in the cache (instead of repos with a work tree and `git pull`) to avoid checking out the whole work tree (which is slow). We don't need the work tree -- just the history.
2. Change the PR description to match what this ended up looking like.

And minor things below.

tgamblin commented:
Oh, and finally: file locking on the caches if that is not already done.

haampie commented:
> This will allow a user to (from the command line) install a spack package at a particular commit

Is this still in the scope of this PR? Or is this about comparison only? Can we get a few lines in the docs about how version comparison works, whether it's necessary to add manual mappings from tags to spack versions or whether they're automatically mapped, what happens if a commit is not before / after a tagged commit (does the latest untagged commit on the release/1.x branch map to version 2.0, `master`, `develop`, infinity?)

becker33 committed:
fix formatting for pytest fixture docstring

becker33 committed:
fix import cycle

becker33 committed:
use bare clones for caching git repos

becker33 committed:
Merge branch 'develop' into add/github-version-comparison

becker33 committed:
flake

becker33 committed:
isort

becker33 committed:
git_url_parse: fix docstring syntax error

becker33 committed:
minor bugfix

haampie commented on lib/spack/spack/version.py:
```diff
@@ -276,10 +319,13 @@ def satisfies(self, other):
         gcc@4.7 so that when a user asks to build with gcc@4.7, we can find
         a suitable compiler.
         """
+        self_cmp = self._cmp(other.commit_lookup)
```
Why was this mixing `self` and `other`?
